[![Abrir en Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/jalonso1979/puremacro/blob/main/curso/notebooks/T00_syllabus_es.ipynb)

> **Nota para Google Colab**: la celda siguiente instala `puremacro` y, si hace falta, descarga del repositorio público [jalonso1979/puremacro](https://github.com/jalonso1979/puremacro) sólo la carpeta `curso/notebooks` (auxiliares, datos congelados y modelos). No pide acceso a Google Drive.

In [1]:
# Configuración del entorno: Google Colab, instalación local o Juno (iPad)
import os
import sys
from pathlib import Path

def _safe_probe(p, is_dir=None):
    """¿Existe la ruta? Sin lanzar OSError/PermissionError en el sandbox de Juno."""
    try:
        p = Path(p)
        if is_dir is True:
            return p.is_dir()
        if is_dir is False:
            return p.is_file()
        return p.exists()
    except (OSError, PermissionError, TypeError, ValueError):
        return False

if "google.colab" in sys.modules:
    !pip install -q "puremacro>=4.0.1,<5" openpyxl
    if not _safe_probe(Path.cwd() / "_nbstyle.py", is_dir=False):
        # Clon parcial del repositorio público: sólo curso/notebooks, sin historia
        _repo = Path("/content/puremacro")
        if not _safe_probe(_repo, is_dir=True):
            !git clone -q --depth 1 --filter=blob:none --sparse https://github.com/jalonso1979/puremacro.git {_repo}
        !git -C {_repo} sparse-checkout set curso/notebooks 2>&1 | tail -n 2
        os.chdir(_repo / "curso/notebooks")

# Carpeta con los auxiliares del curso (_nbstyle, _datos) en sys.path, sin salir del sandbox
_candidatos = [Path.cwd(), Path.cwd() / "curso/notebooks", Path.cwd() / "notebooks"]
try:
    _candidatos.insert(0, Path(__file__).resolve().parent)  # Juno define __file__
except (NameError, OSError):
    pass
for _c in _candidatos:
    if _safe_probe(_c / "_nbstyle.py", is_dir=False):
        if str(_c) not in sys.path:
            sys.path.insert(0, str(_c))
        break

# Gráficas en línea y estilo del curso
%matplotlib inline
try:
    import _nbstyle
    _nbstyle.apply_style()
except ImportError:
    pass

# Macroeconomía Avanzada (MAV) — Syllabus Interactivo

## Manifiesto Académico: Macroeconomía Cuantitativa Moderna

La macroeconomía moderna ha dejado atrás las abstracciones puramente analíticas de lápiz y papel. La economía de agente representativo tipo Robinson Crusoe—aunque fundamental en su momento para comprender la optimización intertemporal—resulta insuficiente para responder a las preguntas estructurales que enfrentan hoy los bancos centrales, los ministerios de hacienda y los organismos multilaterales.

Evaluar el impacto macroeconómico de paradas repentinas (*sudden stops*), crisis de deuda soberana, transiciones de política climática o sorpresas monetarias de alta frecuencia exige marcos que satisfagan tres condiciones indispensables:

1. **Consistencia de Equilibrio General Dinámico**: Hogares, empresas y autoridades económicas optimizan con visión de futuro (*forward-looking*) sujetos a restricciones presupuestarias y de recursos, donde los precios de equilibrio vacían simultáneamente los mercados de bienes, trabajo y activos financieros.
2. **Heterogeneidad Microeconómica y Mercados Incompletos**: La desigualdad de riqueza, el riesgo de ingresos no asegurable, los hogares con restricción de liquidez (*hand-to-mouth*) y las fricciones financieras generan dinámicas agregadas no lineales (capturadas por modelos HANK, *Heterogeneous Agent New Keynesian*) que los modelos de agente representativo ignoran por construcción.
3. **Disciplina Empírica Rigurosa**: Los modelos teóricos no son meros ejercicios matemáticos abstractos; son herramientas de medición estructural calibradas o estimadas frente a la contabilidad nacional, sorpresas financieras de alta frecuencia y microdatos de encuestas de hogares.

Este curso, **Macroeconomía Avanzada (MAV)**, materializa esta frontera. Construido sobre **puremacro**—una biblioteca computacional en Python puro diseñada para ejecutarse localmente o en la nube—cada módulo vincula la teoría matemática dinámica, los algoritmos computacionales y la contrastación empírica con datos reales.

## Arquitectura Metodológica y Dinámica del Curso

En este curso, los modelos económicos operan como laboratorios computacionales. Cada módulo se estructura sobre una secuencia de investigación en cuatro etapas:

$$\text{Teoría Estructural} \;\longrightarrow\; \text{Formulación Computacional} \;\longrightarrow\; \text{Pipeline Empírico (Micro/Macro)} \;\longrightarrow\; \text{Experimentos de Política Contrafactual}$$

- **Teoría**: Planteamiento formal del problema dinámico (ecuaciones de Euler, Bellman, funciones de emparejamiento o restricciones de identificación estructural en VARs).
- **Solución Computacional**: Implementación de algoritmos de solución, desde iteración de la función de valor (VFI) y métodos de perturbación, hasta filtros de Kalman y proyecciones locales.
- **Medición Empírica**: Contraste con cuentas nacionales trimestrales, series de FRED, datos de la OCDE y microdatos laborales (ENOE).
- **Contrafactuales y Pruebas de Estrés**: Sometimiento del modelo a perturbaciones históricas (la Gran Recesión de 2008, el shock de COVID-19 en 2020, o crisis de balanza de pagos) para evaluar disyuntivas de política económica.

In [2]:
import pandas as pd

modules = pd.DataFrame(
    [
        ("1", "Ciclos Económicos y Hechos Estilizados", "Filtros Espectrales y Alta Frecuencia", "T01_A, T01_B, T01_C"),
        ("2", "Crecimiento Neoclásico y Solvers DSGE Locales", "Dynare .mod, Klein QZ, RBC, Nuevo Keynesiano, OccBin", "T02_A, T02_B, T02_C, T02_D"),
        ("3", "Riesgo, Incertidumbre e IA (LLMs)", "GARCH, GAR, NLP Sentimiento", "T03_A a T03_F"),
        ("4", "Real Business Cycles (RBC) y Dashboards", "Simuladores Interactivos", "T04_A, T04_B (app Streamlit)"),
        ("5", "Identificación Empírica (Inflación Post-COVID)", "SVAR, Proyecciones Locales", "T05_A a T05_F"),
        ("6", "Mercados Incompletos y HANK", "Agentes Heterogéneos", "T06_A, T06_B, T06_C"),
        ("7", "Dinámica de Empresas y Mercado de Bienes", "Dinámica de empresas y competencia", "T07_A"),
        ("8", "Flujos Laborales y Fricciones (Microdatos ENOE)", "Matching y Búsqueda", "T08_A"),
        ("9", "Validación, Frontera e Índices Propios", "Econometría Avanzada", "T09_A, T09_B"),
    ],
    columns=["Módulo", "Tema Principal", "Herramientas / Conceptos", "Laboratorios Activos"],
).set_index("Módulo")

# Visualización de la arquitectura curricular
modules

,Tema Principal,Herramientas / Conceptos,Laboratorios Activos
Módulo,,,
1,Ciclos Económicos y Hechos Estilizados,Filtros Espectrales y Alta Frecuencia,"T01_A, T01_B, T01_C"
2,Crecimiento Neoclásico y Solvers DSGE Locales,"Dynare .mod, Klein QZ, RBC, Nuevo Keynesiano, ...","T02_A, T02_B, T02_C, T02_D"
3,"Riesgo, Incertidumbre e IA (LLMs)","GARCH, GAR, NLP Sentimiento",T03_A a T03_F
4,Real Business Cycles (RBC) y Dashboards,Simuladores Interactivos,"T04_A, T04_B (app Streamlit)"
5,Identificación Empírica (Inflación Post-COVID),"SVAR, Proyecciones Locales",T05_A a T05_F
6,Mercados Incompletos y HANK,Agentes Heterogéneos,"T06_A, T06_B, T06_C"
7,Dinámica de Empresas y Mercado de Bienes,Dinámica de empresas y competencia,T07_A
8,Flujos Laborales y Fricciones (Microdatos ENOE),Matching y Búsqueda,T08_A
9,"Validación, Frontera e Índices Propios",Econometría Avanzada,"T09_A, T09_B"


---

# Experimento Macroeconómico 0: Laboratorio de Diagnóstico Computacional

### Motivación Económica: Tendencias, No Estacionariedad y Anatomía de Shocks en Tiempo Real

Antes de abordar los modelos estructurales del Módulo 1, todo macroeconomista cuantitativo debe dominar los cimientos del análisis empírico: **extracción de datos, transformaciones de crecimiento y diagnóstico de estacionariedad**.

En las series de tiempo macroeconómicas agregadas, los datos brutos de cuentas nacionales $Y_t$ (como el Producto Interno Bruto real) presentan tendencias estocásticas no estacionarias con deriva positiva:
$$Y_t = Y_0 \exp(\mu t + \xi_t), \quad \xi_t \sim I(1)$$
Los modelos econométricos estándar y los filtros espectrales (como el filtro de Hodrick-Prescott o los sistemas VAR) requieren series covariacionalmente estacionarias ($I(0)$). Aplicar estos filtros directamente sobre niveles no estacionarios sin verificar su orden de integración genera oscilaciones espurias y correlaciones artificiales (el clásico efecto Slutsky-Yule).

Asimismo, los macroeconomistas analizan el crecimiento económico a diferentes frecuencias:
1. **Tasa de Crecimiento Trimestral Anualizada**:
   $$g_t^q = 400 \times \left( \log Y_t - \log Y_{t-1} \right)$$
   Mide el ritmo de expansión o contracción económica durante un solo trimestre expresado a ritmo anual, amplificando la volatilidad de alta frecuencia.
2. **Tasa de Crecimiento Anual (Interanual)**:
   $$g_t^{yoy} = 100 \times \left( \log Y_t - \log Y_{t-4} \right)$$
   Suaviza el ruido estacional y captura la persistencia tendencial de mediano plazo.

En este experimento introductorio, verificarás tu entorno computacional, evaluarás la asimetría histórica entre la **Gran Crisis Financiera de 2008** y el **shock por COVID-19 de 2020**, y aplicarás la prueba formal de **Augmented Dickey-Fuller (ADF)** para demostrar por qué la macroeconomía moderna exige distinguir entre niveles $I(1)$ y diferencias $I(0)$.

### Tareas Guiadas del Estudiante

1. **Verificación del Pipeline**: Carga la serie de PIB real de Estados Unidos combinando el PIB nominal y el deflactor implícito desde `_datos.gasto()` (o directamente de `GDPC1.csv` en `data_curso/`), calculando la serie encadenada de volumen:
   $$Y_t = \frac{\text{PIB}_t^{\text{nominal}}}{\text{Deflactor}_t} \times 100$$
2. **Transformaciones de Crecimiento**: Calcula el logaritmo en nivel ($100 \log Y_t$), la tasa trimestral anualizada ($g_t^q$) y la tasa interanual ($g_t^{yoy}$).
3. **Prueba de Estrés ante Shocks Históricos**: Identifica el punto mínimo de contracción trimestral anualizada durante la **Gran Crisis Financiera** (2008Q1–2009Q4) frente al colapso por la **pandemia de COVID-19** (2020Q1–2020Q4). Compara cómo la métrica anualizada trimestral amplifica shocks abruptos frente a la tasa interanual.
4. **Diagnóstico Econométrico (Prueba de Raíz Unitaria)**: Ejecuta la prueba de Augmented Dickey-Fuller (Said y Dickey, 1984) contrastando la hipótesis nula $H_0: \gamma = 0$ (raíz unitaria / no estacionariedad) en la especificación:
   $$\Delta y_t = \alpha + \gamma y_{t-1} + \sum_{j=1}^p \delta_j \Delta y_{t-j} + \varepsilon_t$$
   La especificación lleva intercepto pero no tendencia determinista: eso es `regression='c'`, el valor por omisión de `adfuller` y lo que ejecuta el código de abajo.
   Compara el $p$-valor del PIB real en niveles frente al de las tasas de crecimiento.

In [3]:
import inspect
from types import SimpleNamespace

import numpy as np
import pandas as pd
from statsmodels.tsa.stattools import adfuller
import _datos

# ------------------------------------------------------------------------------
# 1. Verificación de Pipeline: Carga de PIB real y transformaciones
# ------------------------------------------------------------------------------
g = _datos.gasto()
pib_usa = g.xs("USA")["gdp"] / g.xs("USA")["gdp_defl"] * 100.0
log_pib = 100.0 * np.log(pib_usa)

# Crecimiento trimestral anualizado: 400 * Delta log(Y_t)
g_q = 400.0 * np.log(pib_usa).diff()

# Crecimiento interanual: 100 * log(Y_t / Y_{t-4})
g_yoy = 100.0 * np.log(pib_usa / pib_usa.shift(4))

# ------------------------------------------------------------------------------
# 2. Comparación de Shocks Históricos: GFC 2008 vs. COVID-19 2020
# ------------------------------------------------------------------------------
gfc_window = g_q.loc["2008-01-01":"2009-12-31"]
cov_window = g_q.loc["2020-01-01":"2020-12-31"]

gfc_min_q = gfc_window.min()
cov_min_q = cov_window.min()

gfc_trough_date = gfc_window.idxmin().to_period("Q")
cov_trough_date = cov_window.idxmin().to_period("Q")

gfc_min_yoy = g_yoy.loc["2008-01-01":"2009-12-31"].min()
cov_min_yoy = g_yoy.loc["2020-01-01":"2020-12-31"].min()

print("=" * 70)
print("ANATOMÍA DE SHOCKS MACROECONÓMICOS: CRISIS 2008 vs. COLAPSO COVID 2020")
print("=" * 70)
print(f"Valle GFC 2008 (Trimestral Anualizado): {gfc_min_q:.2f}% (Trimestre valle: {gfc_trough_date})")
print(f"Valle GFC 2008 (Interanual):            {gfc_min_yoy:.2f}%")
print(f"Valle COVID 2020 (Trimestral Anualizado): {cov_min_q:.2f}% (Trimestre valle: {cov_trough_date})")
print(f"Valle COVID 2020 (Interanual):            {cov_min_yoy:.2f}%")
print("-" * 70)

# ------------------------------------------------------------------------------
# 3. Diagnóstico Econométrico: Prueba de Estacionariedad Augmented Dickey-Fuller
# ------------------------------------------------------------------------------
# statsmodels 0.15 devuelve un objeto de resultados cuando se le llama con
# ``result_object=True``, y si no avisa sobre la tupla heredada; 0.14 --lo que
# sigue sirviendo Colab-- no acepta ese argumento y lanza TypeError.  Se lee la
# firma una sola vez y se le dan a ambas versiones los mismos dos atributos,
# para que el cuaderno abra en cualquiera de las dos.
_ADF_OBJETO = "result_object" in inspect.signature(adfuller).parameters


def adf(serie):
    """Prueba ADF con ``.statistic`` y ``.pvalue`` en statsmodels 0.14 y 0.15."""
    if _ADF_OBJETO:
        return adfuller(serie, result_object=True)
    estadistico, p_valor = adfuller(serie)[:2]
    return SimpleNamespace(statistic=estadistico, pvalue=p_valor)


adf_level = adf(log_pib.dropna())
p_val_level = adf_level.pvalue

adf_diff = adf(g_q.dropna())
p_val_diff = adf_diff.pvalue

print("RESULTADOS DE LA PRUEBA AUGMENTED DICKEY-FULLER (ADF):")
print(f"Log PIB Real (Nivel):        Estadístico = {adf_level.statistic:.3f}, p-valor = {p_val_level:.4f} -> {'Raíz Unitaria (No Estacionario)' if p_val_level > 0.05 else 'Estacionario'}")
print(f"Crecimiento Trimestral (g):  Estadístico = {adf_diff.statistic:.3f}, p-valor = {p_val_diff:.4e} -> {'Raíz Unitaria' if p_val_diff > 0.05 else 'Estacionario (I(0))'}")
print("=" * 70)

ANATOMÍA DE SHOCKS MACROECONÓMICOS: CRISIS 2008 vs. COLAPSO COVID 2020
Valle GFC 2008 (Trimestral Anualizado): -8.85% (Trimestre valle: 2008Q4)
Valle GFC 2008 (Interanual):            -4.06%
Valle COVID 2020 (Trimestral Anualizado): -32.82% (Trimestre valle: 2020Q2)
Valle COVID 2020 (Interanual):            -7.69%
----------------------------------------------------------------------
RESULTADOS DE LA PRUEBA AUGMENTED DICKEY-FULLER (ADF):
Log PIB Real (Nivel):        Estadístico = -1.396, p-valor = 0.5841 -> Raíz Unitaria (No Estacionario)
Crecimiento Trimestral (g):  Estadístico = -12.935, p-valor = 3.6295e-24 -> Estacionario (I(0))


In [4]:
# ------------------------------------------------------------------------------
# Validación y Verificación de Soluciones
# ------------------------------------------------------------------------------
# Verificación de magnitudes históricas de contracción
assert gfc_min_q < -8.0, f"Se esperaba contracción trimestral GFC < -8%, obtenido: {gfc_min_q:.2f}%"
assert cov_min_q < -30.0, f"Se esperaba contracción trimestral COVID < -30%, obtenido: {cov_min_q:.2f}%"
assert gfc_min_yoy < -3.5, f"Se esperaba contracción interanual GFC < -3.5%, obtenido: {gfc_min_yoy:.2f}%"
assert cov_min_yoy < -7.0, f"Se esperaba contracción interanual COVID < -7%, obtenido: {cov_min_yoy:.2f}%"

# Verificación de propiedades de estacionariedad
assert p_val_level > 0.10, f"Se esperaba p-valor en niveles > 0.10 (no estacionario), obtenido: {p_val_level:.4f}"
assert p_val_diff < 0.01, f"Se esperaba p-valor en diferencias < 0.01 (estacionario), obtenido: {p_val_diff:.4e}"

print("✓ Verificación del Experimento 0 Superada: ¡Pipeline de datos, entorno y propiedades econométricas validados correctamente!")

✓ Verificación del Experimento 0 Superada: ¡Pipeline de datos, entorno y propiedades econométricas validados correctamente!


### Interpretación Económica y Enlace al Módulo 1

Los resultados de este laboratorio de diagnóstico establecen dos lecciones empíricas fundamentales para la investigación macroeconómica:

1. **Perfil y Asimetría de los Shocks**: La Gran Crisis Financiera de 2008 fue una recesión clásica de balances originada en el sistema financiero; en su punto más crítico, el ritmo trimestral anualizado cayó $-8.85\%$, mientras que la tasa interanual descendió $-4.06\%$. Por el contrario, la pandemia de COVID-19 en 2020 representó una paralización administrativa abrupta de la actividad económica: la tasa trimestral anualizada se desplomó un $-32.82\%$, una caída casi cuatro veces más profunda que en 2008. Sin embargo, al concentrarse el impacto en el segundo trimestre de 2020, la tasa interanual retrocedió $-7.69\%$.
2. **Orden de Integración y Validez Econométrica**: La prueba ADF no logra rechazar la hipótesis nula de raíz unitaria para el logaritmo del PIB real ($p = 0.5841$), lo que confirma que el nivel de producto es integrado de orden 1 ($I(1)$). Intentar calcular estadísticas de segundo momento o calibrar modelos macroeconómicos directamente sobre niveles no estacionarios conduce a inferencias espurias. En contraste, las primeras diferencias logarítmicas son incuestionablemente estacionarias ($p = 3.63 \times 10^{-24} \ll 0.01$).

Con las bases econométricas y el pipeline computacional verificados, continúa al **Módulo 1** (`T01_A_business_cycle_facts_es`) para descomponer las cuentas nacionales desde el mercado de trabajo y analizar el impacto de los filtros estadísticos sobre la medición de los ciclos económicos.